### PR TITLE
feat(IdRegistry): add transferFor/changeRecoveryAddressFor

### PR DESCRIPTION
**Pull request by horsefacts**
## Motivation

Client apps should be able to sponsor gas for FID transfers and recovery address changes.

## Change Summary

Add signature-based `transferFor` and `changeRecoveryAddressFor` functions to `IdRegistry`.

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [x] The PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] The PR has been tagged with change type label(s) (i.e. documentation, feature, bugfix, or chore)
- [x] The PR's changes adhere to all the requirements in the [contribution guidelines](https://github.com/farcasterxyz/contracts/blob/main/CONTRIBUTING.md#3-proposing-changes)
- [x] All [commits have been signed](https://github.com/farcasterxyz/contracts/blob/main/CONTRIBUTING.md#22-signing-commits)

---

## PR-Codex overview
This PR focuses on adding new functions and tests related to transferring and changing the recovery address for fids in the IdRegistry contract.

### Detailed summary
- Added `changeRecoveryAddressTypehash` function to return the typehash for changing the recovery address.
- Added `changeRecoveryAddressFor` function to allow changing the recovery address for an fid.
- Added `transferFor` function to allow transferring an fid from one address to another using two signed Transfer messages.
- Added test cases for the new functions.

> The following files were skipped due to too many changes: `test/IdRegistry/IdRegistry.t.sol`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/IdRegistry.sol [hunk 1]
```diff
@@ -263,6 +301,33 @@ contract IdRegistry is TrustedCaller, Signatures, Pausable, EIP712, Nonces {
         emit ChangeRecoveryAddress(ownerId, recovery);
     }
 
+    /**
+     * @notice Change the recovery address of the given owner. Caller must provide an
```
Comment (horsefacts) [inline src/IdRegistry.sol:305]:
```suggestion
     * @notice Change the recovery address of fid owned by the owner. Caller must provide an
```

File: src/IdRegistry.sol [hunk 2]
```diff
@@ -263,6 +301,33 @@ contract IdRegistry is TrustedCaller, Signatures, Pausable, EIP712, Nonces {
...
+     *         EIP-712 ChangeRecoveryAddress message signed by the owner.
+     *
+     * @param owner    Owner address of the fid whose recovery address will be changed.
```
Comment (horsefacts) [inline src/IdRegistry.sol:308]:
```suggestion
     * @param owner    Custody address of the fid whose recovery address will be changed.
```

File: test/IdRegistry/IdRegistry.t.sol [hunk 3]
```diff
@@ -703,6 +703,320 @@ contract IdRegistryTest is IdRegistryTestSuite {
...
+        uint256 fromPk,
+        uint256 toPk,
+        uint40 _fromDeadline,
+        uint40 _toDeadline
+    ) public {
+        fromPk = _boundPk(fromPk);
+        toPk = _boundPk(toPk);
+        address from = vm.addr(fromPk);
+        address to = vm.addr(toPk);
+        vm.assume(from != to);
```
[the remark this answers is not in the document]
Comment (horsefacts) [reply test/IdRegistry/IdRegistry.t.sol:961]: Without this assume it will revert with `HasId()`.

File: src/IdRegistry.sol [hunk 4]
```diff
@@ -356,6 +356,47 @@ contract IdRegistry is TrustedCaller, Signatures, Pausable, EIP712, Nonces {
         _unsafeTransfer(fromId, from, to);
     }
 
+    /**
+     * @notice Transfer the fid owned by the from address to another address that does not
+     *         have an fid. Caller must provide two signed Transfer messages: one signed by
+     *         the from address and one signed by the to address.
```
Comment (varunsrin) [inline src/IdRegistry.sol:362]: isn't one of them signed by the recovery address, not the from address?

File: test/IdRegistry/IdRegistry.t.sol [hunk 5]
```diff
@@ -1430,6 +1430,375 @@ contract IdRegistryTest is IdRegistryTestSuite {
...
+        assertEq(idRegistry.idOf(from), 1);
+        assertEq(idRegistry.idOf(to), 0);
+
+        vm.prank(caller);
+        vm.expectRevert(Signatures.InvalidSignature.selector);
+        idRegistry.recoverFor(from, to, recoveryDeadline, recoverySig, toDeadline, toSig);
+
+        assertEq(idRegistry.idCounter(), 1);
+        assertEq(idRegistry.idOf(from), 1);
+        assertEq(idRegistry.idOf(to), 0);
```
Comment (varunsrin) [inline test/IdRegistry/IdRegistry.t.sol:1510]: nit: it might be worth asserting that the recovery address is untouched for all failed cases

**Pull request merged**